jaxtyping/_import_hook.py

```diff
@@ -88,14 +92,71 @@ def _dot_lookup(*elements):
     return out
 
 
-def _str_lookup(string):
-    module = ast.parse(string)
-    (expr,) = module.body
-    return expr.value
+class _Typechecker:
+    lookup = {}
+
+    def __init__(self, typechecker):
+        self.typechecker = typechecker
+
+        hash_string = lambda x: str(hashlib.md5(x.encode("utf-8")).hexdigest())
+
+        if isinstance(typechecker, str):
+            # If the typechecker is a string, we make a decorator ourselves
+            # and pass it to lookup
+            string_to_eval = (
+                "def f(x, *args, **kwargs):\n"
+                + f"  import {typechecker.split('.', 1)[0]}\n"
+                + f"  return {typechecker}(x, *args, **kwargs)"
+            )
+            local_vars = {}
+            exec(string_to_eval, {}, local_vars)
+            self.decorator = local_vars["f"]
+            self.hash = hash_string(typechecker)
+
+        elif callable(typechecker):
+            # If the typechecker is a function, treat it as a decorator
+            self.decorator = typechecker
+            try:
+                self.hash = hash_string(
+                    str(typechecker.__module__) + getsource(typechecker)
```

Comment:
I think we need the `__qualname__` here too, as `getsource` only includes the `__name__`.

Here's an example where two functions have the same `__module__`, `__name__`, and source, but different behaviour:
```python
import inspect

class A:
  x = 2

  def f(self):
    return self.x
  
class B:
  x = 3

  def f(self):
    return self.x

one = A().f
two = B().f
print(one.__module__, inspect.getsource(one))
print(two.__module__, inspect.getsource(two))
```

Reply:
Actually, I've just realised that I think it's worse than this.

Suppose someone has:
```python
def somefunction(x):
    return x

def mydecorator(x):
    return somefunction(x)

with jaxtyping.install_import_hook("foo", mydecorator):
    ...
```
and then runs this, so that the `.pyc` files are generated.

Suppose they then change the contents of `somefunction`, and run it again.

In this case the old `.pyc` files will be hit, because `mydecorator` does not appear to have changed. That's a bug!

Even worse, I think that might be a bug with the current set-up as well. If e.g. `beartype.beartype` changes to some other definition, then again we don't pick up on that difference.

How does Python itself handle cache invalidation?

Reply:
Oh, that's a nice bug. The only solution that comes to my mind (after some googling) is to do something with
```
import dis
def x(y):
     return y + 1
def y(z):
     return x(z + 2)

dis.dis(y)
```
which prints:
```
0 LOAD_GLOBAL              0 (x)
2 LOAD_FAST                0 (z)
4 LOAD_CONST               1 (2)
6 BINARY_ADD
8 CALL_FUNCTION            1
10 RETURN_VALUE
```
But this particular function is a relatively bad fit for our cause, and, to do something smarter, I will have to spend more time digging in the CPython sources.

And, yeah, I have found almost no information about cache handling in Python, so this will also require some time to figure out. But I think a better route is to make custom implementation of ```dis.dis```.

Reply:
What I mean is that while ```dis.dis``` processes only the function body, and all calls are isolated, I think it should be totally possible to detect ```CALL_FUNCTION``` and then replace it with output of the recursive call of ```dis.dis```.

Reply:
Actually, I'm now realising that it maybe isn't as bad as I thought. The modified AST doesn't have a direct `@mydecorator` call. It has a `@jaxtyping._import_hook._Typechecker.lookup[some_hash]` call. We *can* afford to cache the `.pyc` file so long as the value stored in `lookup` is correctly invalidated.

I'd need to think a bit harder about what that entails. I think it means that the current implementation isn't buggy, at least, though. If the AST just has something like `@beartype.beartype` then it will still look up the new version given an old `.pyc` file.

Some interesting links I came across whilst researching this:
- [Optional bytecode cache](https://fedoraproject.org/wiki/Changes/Python_Optional_Bytecode_Cache)
- [PEP 3147](https://peps.python.org/pep-3147)
- [PEP 552](https://peps.python.org/pep-0552/)
- [`importlib.util.source_hash`](https://docs.python.org/3/library/importlib.html#importlib.util.source_hash)

Reply:
So *maybe* what that means is that not only is the current implementation fine, but we could weaken the hash key to just `(typechecker.__module__, typechecker.__qualname__)`? Since that's all that is needed to determine the hash used in the AST, which will then dynamically look up the typechecker at runtime?

For safety I could see us restricting things to global functions only (`isinstance(typechecker, types.FunctionType) and typechecker.__closure__ is None`). I think it would still go wrong given two locally-defined functions, that happen to have the same code but different closed-over variables. (Or maybe we just try to include the closed-over variables in the hash as well? I think that might be difficult, they might be unhashable values, e.g. dictionaries.)

Reply:
Hmm, but why cannot we use locally-defined functions? I think when we do, the expected behavior is defined a bit weirdly. If we typecheck the same module, it is just an awkward decision overall: I don't see why would somebody would do that. And, if we typecheck different modules with different local functions, yes, that is an issue, because we will have only one entry in the ```lookup```, but we can include the module we typecheck in the hash. Then there is no issue, because hashes are different, so the function used by the decorator will be the correct one.


In other words, if we assume that typechecking same module with two different typecheckers never happens, then including the typechecked module name in the hash should work fine, shouldn't it? Also, as you noted, we have the most recent version of the typechecker as the decorator, because it loaded dynamically, so maybe we won't have any issues at all as the result?

Reply:
I don't think we can make that assumption though. People *are* going to do all kinds of crazy things here.

Also, I'd be a bit antsy about including the module name in the hash. Hashes are not injective functions. When we're only hashing a fairly limited collection of things (strings like `"beartype.beartype"`) then I think it's fine; anything more large-scale and ever-so-maybe we get a collision error.

Now that we've seen how complicated this is, I'm thinking we should consider returning to only supporting strings, a la `install_import_hook(..., "beartype.beartype")`. And fix the original issue just by following (a)-(c) in [my original comment](https://github.com/google/jaxtyping/pull/116#issuecomment-1734130592).

Reply:
Yeah, including the typechecked module name in the hash increases the odds of collision quite a bit... And I absolutely agree with your position about increasing chances of collisions: the collision could lead to silently working not as expected, which is extremely bad.

But do you really think that going back to string processing is the best solution we can do? It limits the ability of user too much (I guess), and I can think of tons of ways to introduce bugs, e.g. typos, or, in general, passing any invalid typechecker. 

Limitations of the current approach are indeed annoying, but to me they don't look annoying enough to abandon the idea of allowing the functions. Though, maybe, if we assume that the users are crazy, and we want to be safe from spurious issues about 'why does my typechecker does not work', we should make this approach not the recommended one, but an "advanced" one.

Supporting only strings looks to me like a way to effectively force every user to use either beartype or typeguard: I am usually too scared to change something in a string that is going to be ```eval```-ed, and, overall, I believe that is one of the worst anti-patterns out there.

Reply:
Well, anything invalid should still throw an error. I don't think strings introduce any more risk of a silent bug than in any other case.

Conversely when using functions, I *am* concerned about accidentally failing to invalidate the cache in some edge case. I very much do not want to debug that!

Overall it seems like the wrong balance to have substantially more complicated jaxtyping code, just to make an unusual edge case slightly easier.

(If you really do need a custom typechecker, then it's easy enough to just define it in some other module, and refer to it by string: `install_import_hook(..., "my_module.my_typechecker")`.)

Reply:
Well, I don't want to admit it, but I guess your position is as good as mine: I care more about being "easy to extend" and you care more about "not silently failing". And when thinking about it in this fashion, string-only approach looks significantly better...


Anyways, thanks for clarifying your opinion! I really appreciate that you spent a bit of your time to help me understand why the string-only approach is preferred.

Reply:
No worries! Thanks for being such a good sport. I agree that both positions could reasonably be taken. :)